vivisect/symboliks/common.py

```diff
@@ -507,6 +513,9 @@ def getWidth(self):
         # FIXME should we do something about that?
         return self.kids[1].solve()
 
+    def _reduce(self, emu):
```

Comment:
do we want to return a new Mem object every time?  this will change the _sym_id.  do we care?
same for Call._reduce()

Reply:
we should.  it affects the paradigm, where inconsistency is a nightmare.  i vote for a consistent way to quickly (performance-wise) and easily identifying whether anything has changed.  right now, we don't *really* have that.  we have _sym_id which we can compare before/after... but if LookupVar doesn't change it when updates occur, that's out.  if we're cautious, it may be possible to do such a check without walking the entire tree... but especially for things like walkTree ;)  perhaps not in this PR.  for now, would you mind checking and only returning a new one if something has changed?  is that even a reasonable request (eg.  not going to cause you 16 hours of work)?

Reply:
So I like lists. They gives us something concrete to point to.

Return their original self (modified or not) or don't implement _reduce:
* Var
* Const
* LookupVar
* Arg
* signextension

Return a new instance:
* Mem
* All operators/constraints
* cnot

My changes:
* Call returns a new instance of itself
* Mem reduces it's kids (it already returns a new instance of Mem on master it seems)

Also, looking at it, I don't think LookupVar is right since the reduction of the offset might not persist.

Hmm, so if we didn't have to return a new instance of an object, it would let the solver cache persist, which would be a fortunate side-effect. I do really like keeping work around that we've already done. On the other hand, in the worst case of reducing and check, we would still have to walk the entire AST of effects and compare them. We might be able to cheat a bit there with the solver cache, but if we're aiming for full marks on this, then I don't know if the solver cache would cut it.

Ultimately we'd be trying to compare two arbitrarily sized tree structures on every _reduce. I'm not entirely sure if there's a way to do that with tanking performance too much. But it seems like you have an idea? (` if we're cautious, it may be possible to do such a check without walking the entire tree... but especially for things like walkTree ;)`)